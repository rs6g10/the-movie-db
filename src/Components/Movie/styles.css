.error, .loading {
    margin: 0 auto;
    color: white;
    font-size: larger;;
}

.error {
    color: orangered;;
}