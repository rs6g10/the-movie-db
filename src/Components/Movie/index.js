import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Header from '../Header';
import MovieInfo from '../MovieInfo';
import { parseResponse } from '../../utility';

import './styles.css';

class Movie extends Component {
  constructor(props) {
    super(props);
    this.state = {
      loading: false,
      error: null,
      showSearchResults: true,
    };
  }

  componentDidMount() {
    const movieId = this.props.match.params.movieId;
    const movieUrl = `https://api.themoviedb.org/3/movie/${movieId}?api_key=${this.props.apiKey}`;
    const movieCreditsUrl = `https://api.themoviedb.org/3/movie/${movieId}/credits?api_key=${this.props.apiKey}`;
    const movieVideoUrl = `https://api.themoviedb.org/3/movie/${movieId}/videos?api_key=${this.props.apiKey}`;
    this.fetchMovieDetails(movieUrl);
    this.fetchMovieCredits(movieCreditsUrl);
    this.fetchMovieVideos(movieVideoUrl);
  }

  componentWillReceiveProps(nextProps) {
    this.setState({ showSearchResults: true });
    if (this.props.match.params.movieId !== nextProps.match.params.movieId) {
      const movieId = nextProps.match.params.movieId;
      const movieUrl = `https://api.themoviedb.org/3/movie/${movieId}?api_key=${this.props.apiKey}`;
      const movieCreditsUrl = `https://api.themoviedb.org/3/movie/${movieId}/credits?api_key=${this.props.apiKey}`;
      const movieVideoUrl = `https://api.themoviedb.org/3/movie/${movieId}/videos?api_key=${this.props.apiKey}`;
      this.fetchMovieDetails(movieUrl);
      this.fetchMovieCredits(movieCreditsUrl);
      this.fetchMovieVideos(movieVideoUrl);
    }
  }

  fetchMovieDetails(url) {
    this.setState({ loading: false, error: null });
    fetch(url).then(res => parseResponse(res)).then((datum) => {
      const data = datum.results ? datum.results[0] : datum;
      if (data && data.id) {
        this.setState({
          loading: false,
          movieID: data.id,
          original_title: data.original_title,
          tagline: data.tagline,
          overview: data.overview,
          homepage: data.homepage,
          poster: data.poster_path,
          production: data.production_companies,
          production_countries: data.production_countries,
          genre: data.genres,
          release: data.release_date,
          vote: data.vote_average,
          runtime: data.runtime,
          revenue: data.revenue,
          backdrop: data.backdrop_path,
          showSearchResults: false,
          error: null,
        });
      }
    }).catch((err) => {
      this.setState({ loading: false, error: err });
    });
  }

  fetchMovieCredits(url) {
    fetch(url).then(res => parseResponse(res)).then((datum) => {
      if (datum) {
        this.setState({
          loading: false,
          error: null,
          cast: datum.cast ? datum.cast.splice(0, 5) : [],
          director: datum.crew ?
            datum.crew.filter(d => d.job.toLowerCase() === 'director') :
            [],
        });
      }
    }).catch((err) => {
      this.setState({ loading: false, error: err });
    });
  }

  fetchMovieVideos(url) {
    fetch(url).then(res => parseResponse(res)).then((data) => {
      if (data && data.results && data.results.length > 0) {
        if (data.results[0].site && data.results[0].site.toLocaleLowerCase() === 'youtube') {
          this.setState({
            videoKey: data.results[0].key,
            videoName: data.results[0].name,
          });
        }
      }
    }).catch((err) => {
      this.setState({ loading: false, error: err });
    });
  }

  render() {
    const state = this.state;
    return (
      <div>
        <Header showResults={state.showSearchResults} apiKey={this.props.apiKey} />
        {state.error &&
        <div className="row">
          <span className="error">{state.error.message}</span>
        </div>
        }
        {!state.error && state && <MovieInfo data={state} />}
      </div>
    );
  }
}

Movie.propTypes = {
  apiKey: PropTypes.string.isRequired,
};

export default Movie;
