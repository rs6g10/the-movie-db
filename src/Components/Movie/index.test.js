import '../../test/setupEnzyme';

import React from 'react';
import { shallow } from 'enzyme';

import Movie from './';

describe('Movie', () => {
  it('should render as expected with valid movieId', () => {
    const props = { apiKey: 'blah', match: { params: { movieId: 1234 } } };
    const element = shallow(<Movie {...props} />);

    expect(element).toMatchSnapshot();
  });
});
