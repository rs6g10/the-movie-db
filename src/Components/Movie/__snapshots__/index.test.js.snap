// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie should render as expected with valid movieId 1`] = `
<div>
  <Header
    apiKey="blah"
    showResults={true}
  />
  <MovieInfo
    data={
      Object {
        "error": null,
        "loading": false,
        "showSearchResults": true,
      }
    }
  />
</div>
`;
